From e3f93dda59ec2422f35c57b3a97ec51ebc8d4d7a Mon Sep 17 00:00:00 2001
From: Kang Chen <void0red@hust.edu.cn>
Date: Sun, 16 Apr 2023 09:45:04 +0800
Subject: [PATCH 048/780] [Huawei] mm: oom: move memcg_print_bad_task() out of
 mem_cgroup_scan_tasks()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/src-openeuler/kernel/issues/I6NYW4

--------------------------------

raw call flow:

oom_kill_process
  -> mem_cgroup_scan_tasks(.., .., message)
    -> memcg_print_bad_task(message, ..)

message is "const char*" type, and incorrectly cast to
"oom_control*" type in memcg_print_bad_task.

Fix it by moving memcg_print_bad_task out of mem_cgroup_scan_tasks
and call it in select_bad_process and dump_tasks. Furthermore,
use struct oom_control* directly and remove the useless parm `ret`.

Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Kang Chen <void0red@hust.edu.cn>
(cherry picked from commit 789038c7436e6dc0efb144114992134a52233dbd)

Conflicts:
	include/linux/memcontrol.h
	[Add declaration of memcg_print_bad_task()]

Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 include/linux/memcontrol.h |  8 +++++++-
 mm/memcontrol.c            | 16 +++++++++-------
 mm/oom_kill.c              | 14 ++++++++------
 3 files changed, 24 insertions(+), 14 deletions(-)

diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index ad68c07783de..372f6ec66310 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -412,9 +412,11 @@ DECLARE_STATIC_KEY_FALSE(memcg_qos_stat_key);
 
 bool memcg_low_priority_scan_tasks(int (*)(struct task_struct *, void *),
 				   void *);
-void memcg_print_bad_task(void *arg, int ret);
+void memcg_print_bad_task(struct oom_control *oc);
 extern int sysctl_memcg_qos_handler(struct ctl_table *table,
 		int write, void __user *buffer, size_t *length, loff_t *ppos);
+#else
+void memcg_print_bad_task(struct oom_control *oc);
 #endif
 
 /*
@@ -1695,6 +1697,10 @@ unsigned long mem_cgroup_soft_limit_reclaim(pg_data_t *pgdat, int order,
 {
 	return 0;
 }
+
+static inline void memcg_print_bad_task(struct oom_control *oc)
+{
+}
 #endif /* CONFIG_MEMCG */
 
 /* idx can be of type enum memcg_stat_item or node_stat_item */
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 3df0f6a40bd8..7e83c47cc1ea 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -1210,9 +1210,6 @@ int mem_cgroup_scan_tasks(struct mem_cgroup *memcg,
 			break;
 		}
 	}
-#ifdef CONFIG_MEMCG_QOS
-	memcg_print_bad_task(arg, ret);
-#endif
 	return ret;
 }
 
@@ -3996,14 +3993,12 @@ bool memcg_low_priority_scan_tasks(int (*fn)(struct task_struct *, void *),
 	return oc->chosen ? true : false;
 }
 
-void memcg_print_bad_task(void *arg, int ret)
+void memcg_print_bad_task(struct oom_control *oc)
 {
-	struct oom_control *oc = arg;
-
 	if (!static_branch_likely(&memcg_qos_stat_key))
 		return;
 
-	if (!ret && oc->chosen) {
+	if (oc->chosen) {
 		struct mem_cgroup *memcg;
 
 		memcg = mem_cgroup_from_task(oc->chosen);
@@ -4034,6 +4029,13 @@ int sysctl_memcg_qos_handler(struct ctl_table *table, int write,
 
 	return ret;
 }
+
+#else
+
+void memcg_print_bad_task(struct oom_control *oc)
+{
+}
+
 #endif
 
 #ifdef CONFIG_NUMA
diff --git a/mm/oom_kill.c b/mm/oom_kill.c
index 417ff9574d19..ace93ff828e0 100644
--- a/mm/oom_kill.c
+++ b/mm/oom_kill.c
@@ -408,9 +408,10 @@ static void select_bad_process(struct oom_control *oc)
 {
 	oc->chosen_points = LONG_MIN;
 
-	if (is_memcg_oom(oc))
-		mem_cgroup_scan_tasks(oc->memcg, oom_evaluate_task, oc);
-	else {
+	if (is_memcg_oom(oc)) {
+		if (!mem_cgroup_scan_tasks(oc->memcg, oom_evaluate_task, oc))
+			memcg_print_bad_task(oc);
+	} else {
 		struct task_struct *p;
 
 #ifdef CONFIG_MEMCG_QOS
@@ -473,9 +474,10 @@ static void dump_tasks(struct oom_control *oc)
 	pr_info("Tasks state (memory values in pages):\n");
 	pr_info("[  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name\n");
 
-	if (is_memcg_oom(oc))
-		mem_cgroup_scan_tasks(oc->memcg, dump_task, oc);
-	else {
+	if (is_memcg_oom(oc)) {
+		if (!mem_cgroup_scan_tasks(oc->memcg, dump_task, oc))
+			memcg_print_bad_task(oc);
+	} else {
 		struct task_struct *p;
 
 		rcu_read_lock();
-- 
2.17.1

